### ISSUE TITLE
segmentation fault after exporting LD_PRELOAD

### ISSUE THREAD

**Issue by LiunxPaisley** [TACC/FanStore#1]
Hello, I have some troubles in using FanStore.
I was loading data by running mpiexec in a terminal. It seems to be a daemon process. 
![image](https://user-images.githubusercontent.com/23477633/69925018-25a92900-14e9-11ea-989c-846557b7e79b.png)
Then I set LD_PRELOAD and tried to train alexnet in another terminal. Then I got this fault.
![image](https://user-images.githubusercontent.com/23477633/69925109-6e60e200-14e9-11ea-83f3-d96865c22a51.png)
The OS is Ubuntu 16.04 and I compile FanStore with gcc 5.4. The kernel version is 4.4.0. The CPU is Intel Xeon E5 2680v4.

**Comment by zhaozhang** [comment]
Hi, Thank you for your interest in FanStore. To launch FanStore interactively, I usually run the command, then I have to wait until I see something like below:

```
nDirInfoFileSize = 85667036  nDir = 2002
Rank = 0, copied 20000 files
Rank = 0, copied 40000 files
Rank = 0, copied 60000 files
Rank = 0, copied 80000 files
Rank = 0, copied 100000 files
Rank = 0, copied 120000 files
Rank = 0, copied 140000 files
Rank = 0, copied 160000 files
Rank = 0, copied 180000 files
Rank = 0, copied 200000 files
Rank = 0, copied 220000 files
Rank = 0, copied 240000 files
Rank = 0, copied 260000 files
Rank = 0, copied 280000 files
Rank = 0, copied 300000 files
Rank = 0, copied 320000 files
Finished preparing 1333169 scatter files.
No files will be broadcasted.
Rank = 0: nFileLocal = 333296 nMoreFilesToStore = 0 nMoreFilesToStoreStart= 333296
Finished storing extra 0 files locally.
Ready.
Rank = 1: nFileLocal = 333296 nMoreFilesToStore = 0 nMoreFilesToStoreStart= 666592
Rank = 3: nFileLocal = 333281 nMoreFilesToStore = 0 nMoreFilesToStoreStart= 1333169
Rank = 2: nFileLocal = 333296 nMoreFilesToStore = 0 nMoreFilesToStoreStart= 999888
```
Then you press ctrl+z, then ```bg 1``` to put FanStore process to background. Then the dataset is ready at /tmp/fs_`id -u`. 

Another thing to check is that my imagenet dataset has 2002 directories, while yours shows 2003. You may want to give it a check.

**Comment by LiunxPaisley** [comment]
Thanks for your reply. I fixed the first mistake and I can see the sample files in /tmp/fs_xxxx. But after `export LD_PRELOAD=/home/liupai/software/FanStore/wrapper.so`, I still got segmentation fault with run the command `ls`. 
![image](https://user-images.githubusercontent.com/23477633/69940506-73d81f80-151d-11ea-87c3-1fc11bfc65ac.png)

In addition, I compiled FanStore and met a error in line 910 of read_remote_file.c like `error: call to '__open_missing_mode' declared with attribute error`, then I fixed it by changing the line below in Makefile:
```
gcc -c -DNCX_PTR_SIZE=8 -pipe  -O3  -DLOG_LEVEL=4  -DPAGE_MERGE  read_remote_file.c
```
to 
```
gcc -c -DNCX_PTR_SIZE=8 -pipe  -O0  -DLOG_LEVEL=4  -DPAGE_MERGE  read_remote_file.c
```

**Comment by zhaozhang** [comment]
ok, can you try run "cksum /tmp/fs_1001/train/xxx.jpg"? Replace xxx.jpg
with one real file path.

**Comment by LiunxPaisley** [comment]
If LD_PRELOAD is not set, the command runs normally. But if LD_PRELOAD is set, it comes out the segment fault again.
![image](https://user-images.githubusercontent.com/23477633/69965402-296f9680-154f-11ea-8d7d-eee816e77d8d.png)
I found a server with CentOS and installed FanStore. The commands like `ls` seem to work. I think it may be sensitive to the OS version.
![image](https://user-images.githubusercontent.com/23477633/69965741-dd712180-154f-11ea-813b-60f278ba70b1.png)

**Comment by zhaozhang** [comment]
Yes, probably. Without LD_PRELOAD, it is not running through FanStore. We
did run into the Ubuntu issue before, let me find out.

**Comment by zhaozhang** [comment]
1) Please use "-O0".

If you have sudo privilege, you can temporarily modify
"/usr/include/x86_64-linux-gnu/bits/fcntl2.h" and comment out function
__fortify_function
int open (const char *__path, int __oflag, ...). I will figure out a
solution without changing system include file later.

In makefile, mpicxx -o read_remote_file read_remote_file.o dict.o xxhash.o
ncx_slab.o lzsse8.o -lrt -lpthread

for linking.


2) Ubuntu's library is different from CentOS I used to test on. An updated
wrapper.so can be downloaded at the following link.

https://onedrive.live.com/?authkey=%21APy6bnHdqaS6YGw&cid=60572EB2DB4ABB00&id=60572EB2DB4ABB00%21428&parId=60572EB2DB4ABB00%21427&o=OneUp

**Comment by LiunxPaisley** [comment]
Thanks a lot for your help! I will try this. 👍

**Comment by wiliamhuang** [comment]
Yes. We got issue on Ubuntu before. I found something in my email. 

 I set up my own ubuntu clusters and tested out. It seems working after some tweak. 

1) The suggestion I sent is not correct. "-O0" is necessary to make the code work right. Please use "-O0"

If you have sudo privilege, you can temporarily modify "/usr/include/x86_64-linux-gnu/bits/fcntl2.h" and comment out function __fortify_function int open (const char *__path, int __oflag, ...). I will figure out a solution without changing system include file later. 

In makefile, mpicxx -o read_remote_file read_remote_file.o dict.o xxhash.o ncx_slab.o lzsse8.o -lrt -lpthread

for linking. 

2) Ubuntu's library is different from CentOS I used to test on. An updated wrapper.so can be downloaded at the following link. 

https://onedrive.live.com/?authkey=%21APy6bnHdqaS6YGw&cid=60572EB2DB4ABB00&id=60572EB2DB4ABB00%21428&parId=60572EB2DB4ABB00%21427&o=OneUp

I will find my updated wrapper.c for ubuntu and post it later since I am away from the computer with my latest code. 

-lei

**Comment by nustart0720** [comment]
> Hello, I have some troubles in using FanStore.
> I was loading data by running mpiexec in a terminal. It seems to be a daemon process.
> ![image](https://user-images.githubusercontent.com/23477633/69925018-25a92900-14e9-11ea-989c-846557b7e79b.png)
> Then I set LD_PRELOAD and tried to train alexnet in another terminal. Then I got this fault.
> ![image](https://user-images.githubusercontent.com/23477633/69925109-6e60e200-14e9-11ea-83f3-d96865c22a51.png)
> The OS is Ubuntu 16.04 and I compile FanStore with gcc 5.4. The kernel version is 4.4.0. The CPU is Intel Xeon E5 2680v4.

Have you solved this issue? I failed with the same error though I replaced wrapper.so with the link above.
In my ubuntu 18.04 environment, `ls` works but `ls -alt` and `vim` failed which both shows `Segment fault`.

**Comment by nustart0720** [comment]
When I run like `export LD_PRELOAD=...` and get `Segmentation fault` on Ubuntu 18.04 although using `wrapper.so` you provide above. Choud you give me more hint? thanks in advance! @zhaozhang @wiliamhuang

**Comment by zhaozhang** [comment]
> When I run like `export LD_PRELOAD=...` and get `Segmentation fault` on Ubuntu 18.04 although using `wrapper.so` you provide above. Choud you give me more hint? thanks in advance! @zhaozhang @wiliamhuang

Hi, did you try the solution in this reply(https://github.com/TACC/FanStore/issues/1#issuecomment-560463872)?

**Comment by nustart0720** [comment]
> > When I run like `export LD_PRELOAD=...` and get `Segmentation fault` on Ubuntu 18.04 although using `wrapper.so` you provide above. Choud you give me more hint? thanks in advance! @zhaozhang @wiliamhuang
> 
> Hi, did you try the solution in this reply([#1 (comment)](https://github.com/TACC/FanStore/issues/1#issuecomment-560463872))?

Yes, I have tried the two tweaks you provided, but still failed. :(
On my ubuntu18.04, `read_remote_file` works well in the background but once 'LD_PRELOAD' used, some linux command crushed. I have tried run it on CentOS 7, everything goes well.
<img width="351" alt="image" src="https://user-images.githubusercontent.com/32487463/88992897-10ee9080-d317-11ea-9c2c-389828f862ad.png">
or, Could you give me some hints about which part code of `wrapper.c` or other files can be different  for CentOS 7 and ubuntu18.04? thanks a lot.
